h2o-core/src/main/java/water/webserver/H2OHttpViewImpl.java

```diff
@@ -0,0 +1,194 @@
+  }
+
+  @Override
+  public H2OHttpConfig getConfig() {
+    return config;
+  }
+
+  @Override
+  public Collection<RequestAuthExtension> getAuthExtensions() {
+    return ExtensionManager.getInstance().getAuthExtensions();
+  }
+
+
+  // normal login handler part //todo: consider using mostly the same code as in proxy part below
+
+  @Override
+  public boolean loginHandler(String target, HttpServletRequest request, HttpServletResponse response) throws IOException {
+    if (! isLoginTarget(target)) {
+      return false;
+    }
+
+    if (isPageRequest(request)) {
+      sendLoginForm(request, response);
+    } else {
+      ServletUtils.sendResponseError(response, HttpServletResponse.SC_UNAUTHORIZED, "Access denied. Please login.");
+    }
+    return true;
+  }
+
+  private static void sendLoginForm(HttpServletRequest request, HttpServletResponse response) {
+    final String uri = ServletUtils.getDecodedUri(request);
+    try {
+      byte[] bytes;
+      try (InputStream resource = water.init.JarHash.getResource2("/login.html")) {
+        if (resource == null) {
+          throw new IllegalStateException("Login form not found");
+        }
+        final ByteArrayOutputStream baos = new ByteArrayOutputStream();
+        water.util.FileUtils.copyStream(resource, baos, 2048);
+        bytes = baos.toByteArray();
+      }
+      response.setContentType(RequestServer.MIME_HTML);
+      response.setContentLength(bytes.length);
+      ServletUtils.setResponseStatus(response, HttpServletResponse.SC_OK);
+      final OutputStream os = response.getOutputStream();
+      water.util.FileUtils.copyStream(new ByteArrayInputStream(bytes), os, 2048);
+    } catch (Exception e) {
+      ServletUtils.sendErrorResponse(response, e, uri);
+    } finally {
+      ServletUtils.logRequest("GET", request, response);
+    }
+  }
+
+  private static boolean isPageRequest(HttpServletRequest request) {
+    String accept = request.getHeader("Accept");
+    return (accept != null) && accept.contains(RequestServer.MIME_HTML);
+  }
+
+  private static boolean isLoginTarget(String target) {
+    return target.equals("/login") || target.equals("/loginError");
+  }
+
+  // proxy login handler part
+
+  @Override
+  public boolean proxyLoginHandler(String target, HttpServletRequest request, HttpServletResponse response) throws IOException {
+    if (! isLoginTarget(target)) {
+      return false;
+    }
+
+    if (isPageRequest(request)) {
+      proxySendLoginForm(response);
+    } else {
+      response.sendError(HttpServletResponse.SC_UNAUTHORIZED, "Access denied. Please login.");
+    }
+    return true;
+  }
+
+  private static byte[] proxyLoadLoginFormResource() throws IOException {
+    final InputStream loginFormStream = H2OHttpView.class.getResourceAsStream("/www/login.html");
+    if (loginFormStream == null) {
+      throw new IllegalStateException("Login form resource is missing.");
+    }
+    final ByteArrayOutputStream baos = new ByteArrayOutputStream();
```

Comment:
Close ? Should be closed when this gets out of scope and garbage collected, but we should close right away.

Let's use try-with-resources block please.

Reply:
close() on ByteArrayOutputStream doesn't do anything